Shader debugger improvements

neobrain opened the pull request:
A number of functional improvements to the shader disassembler (which is now an actual debugger):

* When the emulator halts at a VertexLoaded breakpoint, the vertex attributes are shown in the shader debugger
* Unreached instructions are colored yellow in the shader disassembly
* Shaders can be dumped from the GUI
* More output semantics than before are recognized when dumping shaders
* Each shader interpreter cycle records its inputs and outputs now, so that they can be inspected in the GUI (think of this as stepping through the shader and inspecting data, just that it's all done automatically and you just select a particular "iteration" to analyze in the GUI)

Obligatory screenshot:
![Screenshot of shader debugger GUI](http://i.imgur.com/VKuFSkl.png)

This PR pends on #929 .

Subv commented:
Could we get some kind of indication about which attributes came from defaults and which ones came from buffer loaders please?

Subv commented:
Does this allow modifying the shader code on the fly?

neobrain commented:
@Subv Shader modification will not be possible too soon. As for the default attribute suggestion, that sounds like a good idea and I'll check whether it's feasible to implement or not.

Subv commented:
Adding to that, from your screenshot i can't see how one knows how many iterations a particular instruction has been through, does it say that anywhere?

neobrain commented:
@Subv Changing the iteration count selects (i.e. as if the user clicked on it) the instruction processed at that iteration. In the future, the disassembly viewer could be modified to add a small icon next to the "current" instruction. An earlier prototype of the shader debugger just displayed information for the currently selected instruction, but unfortunately that's ambiguous when the instruction might be processed more than once (e.g. in loops or helper functions). Hence the somewhat artificial distinction between "iterations" and "instructions".

Instruction offset refers to the "offset" column of the shader disassembly view. It's the offset of a particular instruction from the start of the shader.

DEST_IN and DEST_OUT account for swizzling and negation.

I'm open for concrete proposals for improvement on the GUI, but for now I consider the added functionality more important than fancy appearance. I'm personally not entirely sure how one would expose the information given in the description text in a nicer way.

Subv commented on src/citra_qt/debugger/graphics_vertex_shader.cpp:
```diff
@@ -17,7 +24,7 @@ using nihstro::Instruction;
 using nihstro::SwizzlePattern;
 
-GraphicsVertexShaderModel::GraphicsVertexShaderModel(QObject* parent): QAbstractItemModel(parent) {
+GraphicsVertexShaderModel::GraphicsVertexShaderModel(GraphicsVertexShaderWidget* parent): QAbstractItemModel(parent), par(parent) {
```
Any reason why the par variable isn't called parent?

neobrain replied:
The base class already has a member function called parent(), and I couldn't find a different good name :/

Subv replied:
any reason why you can't use the member of the base class?

neobrain replied:
That would involve `static_cast` for each use of that member, since the base class only returns a pointer of type `QObject*`.

Subv commented on src/citra_qt/debugger/graphics_vertex_shader.h:
```diff
@@ -43,9 +49,41 @@ private slots:
+
+    void DumpShader();
+
+    /** Reload widget based on the current PICA200 state
```
Comments should begin on the next line

Subv commented on src/citra_qt/debugger/graphics_vertex_shader.h:
```diff
@@ -43,9 +49,41 @@ private slots:
+    GraphicsVertexShaderModel* model;
+
+    // TODO: Move these into a single struct
+    std::array<QLineEdit*, 4*16> input_data;  // A text box for each of the 4 components of up to 16 vertex attributes
```
///

chinhodado commented on src/citra_qt/debugger/graphics_vertex_shader.cpp:
```diff
@@ -34,7 +41,7 @@ int GraphicsVertexShaderModel::columnCount(const QModelIndex& parent) const {
 }
 
 int GraphicsVertexShaderModel::rowCount(const QModelIndex& parent) const {
```
RowCount?

neobrain replied:
This is the implementation of a (Qt) base class method.

Subv replied:
These are overrides from some Qt base class

chinhodado commented on src/citra_qt/debugger/graphics_vertex_shader.cpp:
```diff
@@ -34,7 +41,7 @@ int GraphicsVertexShaderModel::columnCount(const QModelIndex& parent) const {
+    return par->info.code.size();
 }
 
 QVariant GraphicsVertexShaderModel::headerData(int section, Qt::Orientation orientation, int role) const {
```
HeaderData?

neobrain replied:
This is the implementation of a (Qt) base class method.

chinhodado commented on src/citra_qt/debugger/graphics_vertex_shader.cpp:
```diff
@@ -247,53 +266,231 @@ QVariant GraphicsVertexShaderModel::data(const QModelIndex& index, int role) con
 
     auto main_widget = new QWidget;
     auto main_layout = new QVBoxLayout;
     {
```
Why are these blocks needed?

neobrain replied:
They are not needed, but they logically separate the individual sub-layouts from each other, making sure that widgets are attached to the intended sublayout. We use this pattern throughout the Qt code base.

With any alternative method, it might be easy to accidently put a widget into the wrong sublayout.

chinhodado commented on src/citra_qt/debugger/graphics_vertex_shader.h:
```diff
@@ -43,9 +49,41 @@ private slots:
     void OnResumed() override;
 
+    void OnInputAttributeChanged(int index);
+
```
Why is there a new line after each of these?

neobrain replied:
There is no contextual connection between these three functions (contrary to OnBreakPointHit and OnResumed, both of which deal with GPU breakpoints).

neobrain commented:
Formal request for more reviews. If nothing comes up by later today, I'll assume it's fine as is, and will merge it after addressing the outstanding comments.

neobrain commented:
For reference: It might be a good idea to make the input vertex data collapsable in some way, and to display the output vertex data in a similar fashion.

neobrain commented:
Addressed comments (the new patches don't need squashing), and rebased on latest master.

QToolBox doesn't support the UI I had in mind, hence none of the contents are expandable, unfortunately.
